File: sdk/nodejs/src/impl/repository-impl.ts
import {Client} from '../client';
import {Entity} from "../entity";
import {Repository} from "../repository";
import {EntityInfo} from "../entity-info";
import {Container} from "../container";
import {BooleanExpression, Event} from "../model/extension";
import {Watcher} from "../watcher";
import {WatcherImpl} from "./watcher-impl";
import {ListRecordParams} from "../list-record-params";

export class RepositoryImpl<T extends Entity> implements Repository<T> {

    public constructor(private client: Client, private entityInfo: EntityInfo) {
    }

    public create(record: T): Promise<T> {
        return this.client.createRecord<T>(this.entityInfo, record);
    }

    public get(id: string, resolveReferences?: string[] ): Promise<T> {
        return this.client.getRecord<T>(this.entityInfo, {id, resolveReferences});
    }

    public update(record: Partial<T> & Entity): Promise<T> {
        return this.client.updateRecord<T>(this.entityInfo, record);
    }

    public delete(id: string): Promise<T> {
        return this.client.deleteRecord<T>(this.entityInfo, id);
    }

    public apply(record: Partial<T>): Promise<T> {
        return this.client.applyRecord<T>(this.entityInfo, record);
    }

    public load(record: Partial<T>, resolveReferences?: string[]): Promise<T> {
        return this.client.loadRecord<T>(this.entityInfo, record, resolveReferences);
    }

    public list(params: ListRecordParams = {}): Promise<Container<T>> {
        return this.client.listRecords<T>(this.entityInfo, params);
    }

    public watch(eventConsumer: (event: Event) => void): Watcher<T> {
        return new WatcherImpl<T>(this.client, this.entityInfo, eventConsumer);
    }
}

File: sdk/java/src/main/java/io/apibrew/client/impl/RepositoryImpl.java
package io.apibrew.client.impl;

import io.apibrew.client.*;
import io.apibrew.client.model.Extension;

import java.util.List;
import java.util.Map;
import java.util.function.Consumer;

public class RepositoryImpl<T extends Entity> implements Repository<T> {
    private final Client client;
    private final EntityInfo<T> entityInfo;

    public RepositoryImpl(ClientImpl client, Class<T> entityClass) {
        this.client = client;
        this.entityInfo = EntityInfo.fromEntityClass(entityClass);
    }

    public RepositoryImpl(ClientImpl client, EntityInfo<T> entityInfo) {
        this.client = client;
        this.entityInfo = entityInfo;
    }

    @Override
    public T create(T record) {
        return client.createRecord(entityInfo, record);
    }

    @Override
    public T get(String id) {
        return client.getRecord(entityInfo, GetRecordParams.builder().id(id).build());
    }

    @Override
    public T get(String id, List<String> resolveReferences) {
        return client.getRecord(entityInfo, GetRecordParams.builder().id(id).resolveReferences(resolveReferences).build());
    }

    @Override
    public T findBy(Map<String, String> filters, List<String> resolveReferences) {
        Container<T> result = list(ListRecordParams.builder()
                .filters(filters)
                .limit(1)
                .resolveReferences(resolveReferences)
                .build());

        if (result.getTotal() > 1) {
            throw new ApiException(Extension.Code.RECORD_VALIDATION_ERROR, "More than one record found");
        } else if (result.getTotal() == 1) {
            return result.getContent().get(0);
        } else {
            throw new ApiException(Extension.Code.RECORD_NOT_FOUND, "Record not found");
        }
    }

    @Override
    public T update(T record) {
        return client.updateRecord(entityInfo, record);
    }

    @Override
    public T delete(String id) {
        return client.deleteRecord(entityInfo, id);
    }

    @Override
    public T apply(T record) {
        return client.applyRecord(entityInfo, record);
    }

    @Override
    public Container<T> list() {
        return client.listRecords(entityInfo, null);
    }

    @Override
    public Container<T> list(ListRecordParams params) {
        return client.listRecords(entityInfo, params);
    }

    @Override
    public T load(T record) {
        return client.loadRecord(entityInfo, record);
    }

    @Override
    public Container<T> list(Extension.BooleanExpression... query) {
        if (query.length == 0) {
            return client.listRecords(entityInfo, null);
        } else if (query.length == 1) {
            return client.listRecords(entityInfo, ListRecordParams.builder().query(query[0]).build());
        } else {
            return client.listRecords(entityInfo, ListRecordParams.builder().query(BooleanExpressionBuilder.and(query)).build());
        }
    }

    @Override
    public Watcher<T> watch(Consumer<Extension.Event> eventConsumer) {
        return new WatcherImpl<>(client, entityInfo, eventConsumer);
    }
}
